---
title: "Assigning with DADA2"
author: "Liangzi"
date: "2025-03-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                       fig.align = "center",
                      # Send figures generated in this file to this folder below
                      fig.path = "../figures/02_AssignASVs/")
```

# Goals

1. Infer Errors in our sequences, separately on forward and reverse reads.
2. Assign ASVs on both forward and reverse reads separately. Apply the error model.
3. Merge forward and reverse ASVs into "contiguous ASVs".
4. Generate first draft of ASV count table.
5. Quality Trimming ASV length.
6. Remove chimeras.
7. Assign Taxonomy with Silva Database.
8. Write out relevant files: 'asv_table', 'asv_fastas', 'tax_table', and 'sample_data'.

## Input
1. Filtered fastq.gz files generated from '01_QualityTrimming.Rmd'.
2. Sample name vector.

## Output
1. 'asv_table'
2. 'asv_fastas'
3. 'tax_table'
4. 'sample_data'

# Set up the encironment

## Set Seed
```{r set-seed}
set.seed(238428)
```

## Load Packages
```{r load-packages}
pacman::p_load(tidyverse, devtools, dada2, 
               patchwork, DT, install = FALSE)
```

# Load Filtered Fastq Files
```{r load-filtered-fastqs}
# Place filtered seq into a variable
filtered_fastqs_path <- "data/01_DADA2/02_filtered_fastqs"

# Intuition Check 
filtered_fastqs_path

# create Forward vector
filtered_forward_reads <- 
  list.files(filtered_fastqs_path, pattern = "R1_filtered.fastq.gz",
             full.names = TRUE)
# Check 
filtered_forward_reads[1:5]

# Reverse vector
filtered_reverse_reads <- 
    list.files(filtered_fastqs_path, pattern = "R2_filtered.fastq.gz",
             full.names = TRUE)  
# Check 
filtered_reverse_reads[1:5]
```

# Sample Name
```{r sample-name}
# Create vector of sample names from the filenames 
sample_names <- sapply(strsplit(basename(filtered_forward_reads), "_"), `[`,1) 

# Intuition Check 
head(sample_names)
```

# Error Modelling
```{r learn-errors}
# Forward Reads
error_forward_reads <-
  learnErrors(filtered_forward_reads, multithread = 6)
# Plot
forward_error_plot <- 
  plotErrors(error_forward_reads, nominalQ = TRUE) + 
  labs(title = "Forward Read Error Model")

# Reverse Reads


```


